include/picongpu/plugins/xrayScattering/XrayScattering.hpp

```diff
@@ -0,0 +1,842 @@
+
+        // memory:
+        using ComplexBuffer = GridBuffer< complex_X, DIM1 >;
+        std::unique_ptr< ComplexBuffer > amplitude;
+        // Needed as long as opePMD-api doesn't support complex values:
+        //! Storage for amplitude real part used when dumping data
+        std::vector< float_X > realPart;
+        //! Storage for amplitude imaginary part used when dumping data
+        std::vector< float_X > imgPart;
+        // Used only in the split mode:
+        //! Storage for receiving amplitude data from another node
+        std::vector< complex_X > amplitudeReceive;
+        //! Number of scattering vectors on initialy last rank
+        uint64_t  resOfVectors;
+        // Used only in the mirrored mode:
+        std::vector< complex_X > amplitudeMaster;
+
+        // Variables for plugin options:
+        std::string notifyPeriod;
+        std::string speciesName;
+        std::string pluginName;
+        std::string pluginPrefix;
+        std::string fileName;
+        std::string fileExtension;
+        std::string compressionMethod;
+        std::string outputPeriod_s;
+        std::string memoryLayout;
+        //! Plugin functioning mode
+        OutputMemoryLayout outputLayout;
+        //! Time steps at which the output is dumped
+        using SeqOfTimeSlices = std::vector< pluginSystem::TimeSlice >;
+        SeqOfTimeSlices outputPeriod;
+
+        /** Range of scattering vector
+         * The scattering vector here is defined as
+         * 4*pi*sin(theta)/lambda, where 2 * theta is the angle between the
+         * incoming k-vector and the scattered one.
+         * See the definition in this paper https://doi.org/10.1063/1.5008289.
+         **/
+        float2_X q_min, q_max, q_step;
+        //! Number of scattering vectors
+        DataSpace< DIM2 > numVectors;
+
+        uint32_t totalSimulationCells;
+
+        // Needed to handle the parallelization over multiple hosts.
+        bool isMaster;
+        uint32_t mpiRank;
+        //! Total number of nodes
+        uint32_t countRanks;
+        //! Number of Times the split output was passed along
+        uint32_t accumulatedRotations;
+        mpi::MPIReduce reduce;
+
+        //! Output writer
+        std::unique_ptr< XrayScatteringWriter< float_X > > dataWriter;
+
+
+    public:
+
+        //! XrayScattering object initializer.
+        XrayScattering( )
+            : pluginName(
+            "xrayScattering: Calculate the SAXS scattering intensity of a "
+            "species."
+        ),
+            speciesName( T_ParticlesType::FrameType::getName( ) ),
+            pluginPrefix(speciesName + std::string( "_xrayScattering" ) ),
+            // this is bodged so it passes the verification at
+            // MappingDescription.hpp:79
+            cellDescription( DataSpace< simDim >( SuperCellSize::toRT( ) ) ),
+            isMaster( false ),
+            currentStep( 0 ),
+            gc( Environment< simDim >::get( ).GridController( ) ),
```

Comment:
I think it is unsafe to use this, or other singletones, from this constructor. The problem is that this code may be called before other initialization is finished, which may break the initialization order. This is what happens on my machine when run locally. Using `Environment< simDim >::get( ).GridController( )` from inside `notify()` and other "run-time" methods is fine.

Reply:
Do you think I can I run it from `pluginLoad( )`? Or should I run it every time I use a `GridController`?

Reply:
You can, but not sure if that helps. Currently you have a reference as a class member and that can only be initialized in the constructor.

Reply:
What is the status of this one? appears unchanged unless i am missing something

Reply:
well this is weird,... I was changing this. Let me look what happend

Reply:
@sbastrakov  This 41151b0462783d6379440a1dc41107eb37d286fa  has changed it but somehow didn't ? I'm confused :laughing:

Reply:
Oh ok you have to make sure that all commits are chosen when viewing changes. If you just click on view changes next to this post it's going to show you the stage from the time this was posted.

Reply:
Also I have left this in `pluginLoad( )`. I use it quite a lot there and I didn't now what else to do.

Reply:
Ah, sorry, I was indeed looking only at changes of some commits, but not all. Thanks for explaining. Yes, using from `pluginLoad()` is fine.